File: Python/Problems/Problems/problem2909.py
import heapq
import math
import sys

from utils import print1DArray, print2DArray
from typing import List, Set, Optional
from collections import Counter

from .problem import Problem


class Problem2909(Problem):
    def test(self) -> bool:
        nums = [8, 6, 1, 5, 3]

        expected = 9

        result = self.minimumSum(nums)

        print(result)

        return result == expected

    def minimumSum(self, nums: List[int]) -> int:
        n = len(nums)
        rightMins = [0] * n
        rightMins[n - 1] = nums[n - 1]
        for i in range(n - 2, -1, -1):
            rightMins[i] = min(nums[i], rightMins[i + 1])
        curMin = nums[0]
        result = sys.maxsize
        for i in range(n - 1):
            if nums[i] > curMin and nums[i] > rightMins[i + 1]:
                result = min(result, curMin + nums[i] + rightMins[i + 1])
            curMin = min(curMin, nums[i])

        return -1 if result == sys.maxsize else result

File: C++/Problems/Problems/problem_2909.cpp
#include "problem_2909.h"

bool problem_2909::test() {
    std::vector nums { 8, 6, 1, 5, 3 };

    int expected = 9;

    auto result = minimumSum(nums);

    std::cout << result << std::endl;

    return result == expected;
}

int problem_2909::minimumSum(std::vector<int> &nums) {
    int n = nums.size();
    std::vector<int> rightMins(n);
    rightMins[n - 1] = nums[n - 1];
    for (int i = n - 2; i > 0; --i)
    {
        rightMins[i] = std::min(nums[i], rightMins[i + 1]);
    }
    int curMin = nums[0];
    int result = INT_MAX;
    for (int i = 1; i < n - 1; ++i)
    {
        if (nums[i] > curMin && nums[i] > rightMins[i + 1])
            result = std::min(result, curMin + nums[i] + rightMins[i + 1]);
        curMin = std::min(curMin, nums[i]);
    }

    return result == INT_MAX ? -1 : result;
}
